File: python/code_reviewer/src/code_reviewer/tools.py
import os
import re

import httpx
from agnt5 import Context, FunctionContext, tool
from agnt5._ids import generate_cid

from code_reviewer.functions import (
    call_jira_api,
    call_linear_api,
    call_github_api,
)

from code_reviewer.utils import parse_adf


def _make_fun_ctx(ctx: Context) -> FunctionContext:
    """Create a FunctionContext from an agent/workflow Context for calling @function helpers."""
    return FunctionContext(
        run_id=ctx.run_id,
        correlation_id=generate_cid(),
        parent_correlation_id=getattr(ctx, "_correlation_id", ""),
        runtime_context=getattr(ctx, "_runtime_context", None),
        trace_metadata=getattr(ctx, "_trace_metadata", None),
    )


@tool
async def jira_ticket_fetcher(ctx: Context, ticket_url: str) -> dict:
    """
    Fetch Jira issue details via REST API (v3).

    Args:
        ctx: FunctionContext for logging
        ticket_url: Full Jira ticket URL
            (e.g., https://company.atlassian.net/browse/PROJ-123)

    Returns:
        dict: Normalized ticket data with keys: key, summary, status,
            assignee, priority, project, description, url

    Raises:
        EnvironmentError: If Jira credentials are missing
        ValueError: If ticket URL is invalid
    """
    jira_email = os.getenv("JIRA_EMAIL")
    jira_token = os.getenv("JIRA_API_TOKEN")
    jira_domain = os.getenv("JIRA_DOMAIN")

    if not all([jira_email, jira_token, jira_domain]):
        error_msg = ("❌ Missing Jira credentials "
                     "(JIRA_EMAIL, JIRA_API_TOKEN, JIRA_DOMAIN)")
        ctx.logger.error(error_msg)
        raise EnvironmentError("Missing Jira credentials in environment.")

    match = re.search(r"/browse/([A-Z0-9\-]+)", ticket_url)
    if not match:
        ctx.logger.error(f"❌ Invalid Jira ticket URL: {ticket_url}")
        raise ValueError(f"Invalid Jira ticket URL: {ticket_url}")

    ticket_key = match.group(1)
    api_url = f"{jira_domain}/rest/api/3/issue/{ticket_key}"

    ctx.logger.info(f"🚀 Fetching Jira issue {ticket_key} from {api_url}")
    fun_ctx = _make_fun_ctx(ctx)
    data = await call_jira_api(
        fun_ctx, url=api_url, auth=(jira_email, jira_token)
    )
    fields = data.get("fields", {})

    description_text = parse_adf(fields.get("description", {})).strip()

    result = {
        "key": data.get("key"),
        "summary": fields.get("summary"),
        "status": fields.get("status", {}).get("name"),
        "assignee": (
            fields.get("assignee", {}).get("displayName") if fields.get("assignee") else None
        ),
        "priority": fields.get("priority", {}).get("name"),
        "project": fields.get("project", {}).get("name"),
        "description": description_text or None,
        "url": ticket_url,
    }

    ctx.logger.info(f"✅ Jira issue {ticket_key} fetched successfully.")
    return result


@tool
async def linear_ticket_fetcher(ctx: Context, ticket_url: str) -> dict:
    """
    Fetch Linear issue details using GraphQL API.

    Args:
        ctx: FunctionContext for logging
        ticket_url: Full Linear ticket URL
            (e.g., https://linear.app/team/issue/PROJ-123)

    Returns:
        dict: Normalized ticket data with keys: key, summary, status,
            assignee, priority, project, description, url

    Raises:
        EnvironmentError: If LINEAR_API_TOKEN is missing
        ValueError: If ticket URL is invalid or issue not found
    """
    linear_token = os.getenv("LINEAR_API_TOKEN")

    if not linear_token:
        ctx.logger.error("❌ Missing LINEAR_API_TOKEN")
        raise EnvironmentError("Missing LINEAR_API_TOKEN in environment variables.")

    match = re.search(r"/issue/([A-Z0-9\-]+)", ticket_url)
    if not match:
        ctx.logger.error(f"❌ Invalid Linear ticket URL: {ticket_url}")
        raise ValueError(f"Invalid Linear ticket URL: {ticket_url}")

    ticket_key = match.group(1)
    ctx.logger.info(f"🚀 Fetching Linear issue {ticket_key}")

    fun_ctx = _make_fun_ctx(ctx)
    data = await call_linear_api(
        fun_ctx, ticket_id=ticket_key, linear_token=linear_token
    )
    issue = data.get("data", {}).get("issue")

    if not issue:
        ctx.logger.error(f"❌ No issue found for Linear key: {ticket_key}")
        raise ValueError(f"No issue found for Linear key: {ticket_key}")

    result = {
        "key": issue["identifier"],
        "summary": issue["title"],
        "status": issue.get("state", {}).get("name"),
        "assignee": issue.get("assignee", {}).get("name"),
        "priority": str(issue.get("priority")) if issue.get("priority") else None,
        "project": issue.get("team", {}).get("name"),
        "description": issue.get("description"),
        "url": issue["url"],
    }

    ctx.logger.info(f"✅ Linear issue {ticket_key} fetched successfully.")
    return result


@tool
async def pr_fetcher(ctx: Context, pr_url: str) -> dict:
    """
    Fetch Pull Request metadata and ALL file diffs from GitHub REST API.

    Paginates through all changed files so no file is missed.

    Args:
        ctx: FunctionContext for logging
        pr_url: Full GitHub PR URL
            (e.g., https://github.com/owner/repo/pull/123)

    Returns:
        dict: PR data with keys: repo, pr_number, title, author, state,
            created_at, merged_at, description, changed_files, additions,
            deletions, files (full patches, all pages)

    Raises:
        EnvironmentError: If GITHUB_TOKEN is missing
        ValueError: If PR URL is invalid
    """
    token = os.getenv("GITHUB_TOKEN")

    if not token:
        ctx.logger.error("❌ Missing GITHUB_TOKEN")
        raise EnvironmentError("Missing GITHUB_TOKEN in environment variables.")

    match = re.match(r"https://github.com/([^/]+)/([^/]+)/pull/(\d+)", pr_url)
    if not match:
        ctx.logger.error(f"❌ Invalid PR URL: {pr_url}")
        raise ValueError(f"Invalid PR URL: {pr_url}")

    owner, repo, pr_number = match.groups()
    api_url = f"https://api.github.com/repos/{owner}/{repo}/pulls/{pr_number}"
    headers = {
        "Authorization": f"token {token}",
        "Accept": "application/vnd.github.v3+json",
    }

    ctx.logger.info(f"🚀 Fetching PR #{pr_number} from {owner}/{repo}")

    fun_ctx = _make_fun_ctx(ctx)
    pr_meta = await call_github_api(fun_ctx, url=api_url, headers=headers)

    # Paginate through ALL changed files (GitHub returns 30/page by default)
    files: list = []
    files_url = f"https://api.github.com/repos/{owner}/{repo}/pulls/{pr_number}/files"
    async with httpx.AsyncClient(headers=headers, timeout=30.0) as client:
        page_url = f"{files_url}?per_page=100&page=1"
        while page_url:
            ctx.logger.info(f"📄 Fetching files page: {page_url}")
            resp = await client.get(page_url)
            resp.raise_for_status()
            files.extend(resp.json())
            link_header = resp.headers.get("Link", "")
            next_url = None
            for part in link_header.split(","):
                part = part.strip()
                if 'rel="next"' in part:
                    next_url = part.split(";")[0].strip().strip("<>")
                    break
            page_url = next_url

    result = {
        "repo": f"{owner}/{repo}",
        "pr_number": int(pr_number),
        "title": pr_meta.get("title"),
        "author": pr_meta.get("user", {}).get("login"),
        "state": pr_meta.get("state"),
        "created_at": pr_meta.get("created_at"),
        "merged_at": pr_meta.get("merged_at"),
        "description": pr_meta.get("body"),
        "changed_files": pr_meta.get("changed_files"),
        "additions": pr_meta.get("additions"),
        "deletions": pr_meta.get("deletions"),
        "files": [
            {
                "filename": f.get("filename"),
                "status": f.get("status"),
                "additions": f.get("additions"),
                "deletions": f.get("deletions"),
                "changes": f.get("changes"),
                "patch": f.get("patch", "") or "",
                "has_patch": bool(f.get("patch")),
            }
            for f in files
        ],
    }

    ctx.logger.info(
        f"✅ PR #{pr_number} fetched: {len(files)} files across all pages, "
        f"+{pr_meta.get('additions', 0)} -{pr_meta.get('deletions', 0)}"
    )
    return result


@tool
async def detect_ticket_source(ctx: Context, ticket_url: str) -> str:
    """
    Determine ticketing platform (Jira or Linear) based on URL pattern.

    Args:
        ctx: FunctionContext for logging
        ticket_url: Ticket URL to analyze

    Returns:
        str: Tool name - "jira" or "linear"

    Raises:
        ValueError: If URL doesn't match known patterns
    """
    if "atlassian.net" in ticket_url:
        ctx.logger.info("🟦 Detected Jira ticket URL")
        return "jira"
    elif "linear.app" in ticket_url:
        ctx.logger.info("🟪 Detected Linear ticket URL")
        return "linear"
    else:
        ctx.logger.error(f"❌ Unsupported ticket URL: {ticket_url}")
        raise ValueError(f"Unsupported ticketing platform URL: {ticket_url}")


__all__ = [
    "jira_ticket_fetcher",
    "linear_ticket_fetcher",
    "pr_fetcher",
    "detect_ticket_source",
]

File: go/code_reviewer/src/code_reviewer/tools.go
// Agent tools: PR fetch, Jira/Linear ticket fetch, and ticket-source
// detection, wrapping GitHub REST, Jira REST v3, and Linear GraphQL.
package code_reviewer

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"regexp"
	"strconv"
	"strings"
	"time"

	"agnt5.dev/sdk-go/agnt5"
)

var apiClient = &http.Client{Timeout: 30 * time.Second}

var (
	prURLPattern     = regexp.MustCompile(`^https://github\.com/([^/]+)/([^/]+)/pull/(\d+)`)
	jiraKeyPattern   = regexp.MustCompile(`/browse/([A-Z0-9\-]+)`)
	linearKeyPattern = regexp.MustCompile(`/issue/([A-Z0-9\-]+)`)
)

func doJSON(req *http.Request, out any) error {
	resp, err := apiClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(body))
	}
	return json.NewDecoder(resp.Body).Decode(out)
}

// callGitHubAPI fetches one GitHub API URL with retry (matches the Python
// template's call_github_api function, which has its own RetryPolicy).
func callGitHubAPI(ctx context.Context, url, token string, out any) error {
	return retryWithBackoff(ctx, 3, func() error {
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
		if err != nil {
			return err
		}
		req.Header.Set("Authorization", "token "+token)
		req.Header.Set("Accept", "application/vnd.github.v3+json")
		return doJSON(req, out)
	})
}

// fetchAllPRFiles paginates through every changed file on a PR (GitHub caps
// at 30/page by default), following the Link response header.
func fetchAllPRFiles(ctx context.Context, filesURL, token string) ([]PRFile, error) {
	var allFiles []PRFile
	pageURL := filesURL + "?per_page=100&page=1"

	for pageURL != "" {
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, pageURL, nil)
		if err != nil {
			return nil, err
		}
		req.Header.Set("Authorization", "token "+token)
		req.Header.Set("Accept", "application/vnd.github.v3+json")

		var resp *http.Response
		err = retryWithBackoff(ctx, 3, func() error {
			var doErr error
			resp, doErr = apiClient.Do(req)
			return doErr
		})
		if err != nil {
			return nil, err
		}

		var page []struct {
			Filename  string `json:"filename"`
			Status    string `json:"status"`
			Additions int    `json:"additions"`
			Deletions int    `json:"deletions"`
			Patch     string `json:"patch"`
		}
		decodeErr := json.NewDecoder(resp.Body).Decode(&page)
		linkHeader := resp.Header.Get("Link")
		resp.Body.Close()
		if decodeErr != nil {
			return nil, decodeErr
		}

		for _, f := range page {
			allFiles = append(allFiles, PRFile{
				Filename: f.Filename, Status: f.Status,
				Additions: f.Additions, Deletions: f.Deletions,
				Patch: f.Patch, HasPatch: f.Patch != "",
			})
		}

		pageURL = ""
		for _, part := range strings.Split(linkHeader, ",") {
			part = strings.TrimSpace(part)
			if strings.Contains(part, `rel="next"`) {
				if i := strings.Index(part, ";"); i > 0 {
					pageURL = strings.Trim(strings.TrimSpace(part[:i]), "<>")
				}
				break
			}
		}
	}
	return allFiles, nil
}

// fetchPR fetches PR metadata and ALL changed files (paginated) from the
// GitHub REST API.
func fetchPR(ctx context.Context, prURL, githubToken string) (PRData, error) {
	match := prURLPattern.FindStringSubmatch(prURL)
	if match == nil {
		return PRData{}, fmt.Errorf("invalid PR URL: %s", prURL)
	}
	owner, repo, prNumberStr := match[1], match[2], match[3]
	prNumber, _ := strconv.Atoi(prNumberStr)

	apiURL := fmt.Sprintf("https://api.github.com/repos/%s/%s/pulls/%s", owner, repo, prNumberStr)
	var meta struct {
		Title string `json:"title"`
		User  struct {
			Login string `json:"login"`
		} `json:"user"`
		State        string `json:"state"`
		Body         string `json:"body"`
		ChangedFiles int    `json:"changed_files"`
		Additions    int    `json:"additions"`
		Deletions    int    `json:"deletions"`
	}
	if err := callGitHubAPI(ctx, apiURL, githubToken, &meta); err != nil {
		return PRData{}, err
	}

	filesURL := fmt.Sprintf("https://api.github.com/repos/%s/%s/pulls/%s/files", owner, repo, prNumberStr)
	files, err := fetchAllPRFiles(ctx, filesURL, githubToken)
	if err != nil {
		return PRData{}, err
	}

	return PRData{
		Repo: owner + "/" + repo, PRNumber: prNumber,
		Title: meta.Title, Author: meta.User.Login, State: meta.State,
		Description: meta.Body, ChangedFiles: meta.ChangedFiles,
		Additions: meta.Additions, Deletions: meta.Deletions, Files: files,
	}, nil
}

func NewPRFetcherTool(cfg AppConfig) (agnt5.Tool, error) {
	return agnt5.NewTool("pr_fetcher", func(c context.Context, args map[string]any) (any, error) {
		prURL, _ := args["pr_url"].(string)
		logInfo(c, "Fetching PR", "url", prURL)
		data, err := fetchPR(c, prURL, cfg.GitHubToken)
		if err != nil {
			logError(c, "PR fetch failed", "error", err)
			return nil, err
		}
		logInfo(c, "PR fetched", "files", len(data.Files), "additions", data.Additions, "deletions", data.Deletions)
		return data, nil
	},
		agnt5.WithToolDescription("Fetch Pull Request metadata and ALL file diffs from GitHub REST API."),
		agnt5.WithToolSchema(map[string]any{
			"type": "object",
			"properties": map[string]any{
				"pr_url": map[string]any{"type": "string", "description": "Full GitHub PR URL (e.g., https://github.com/owner/repo/pull/123)"},
			},
			"required": []string{"pr_url"},
		}),
	)
}

func NewJiraTicketFetcherTool(cfg AppConfig) (agnt5.Tool, error) {
	return agnt5.NewTool("jira_ticket_fetcher", func(c context.Context, args map[string]any) (any, error) {
		ticketURL, _ := args["ticket_url"].(string)
		data, err := fetchJiraTicket(c, ticketURL, cfg)
		if err != nil {
			logError(c, "Jira fetch failed", "error", err)
			return nil, err
		}
		return data, nil
	},
		agnt5.WithToolDescription("Fetch Jira issue details via REST API (v3)."),
		agnt5.WithToolSchema(map[string]any{
			"type": "object",
			"properties": map[string]any{
				"ticket_url": map[string]any{"type": "string", "description": "Full Jira ticket URL (e.g., https://company.atlassian.net/browse/PROJ-123)"},
			},
			"required": []string{"ticket_url"},
		}),
	)
}

func fetchJiraTicket(ctx context.Context, ticketURL string, cfg AppConfig) (TicketData, error) {
	if cfg.JiraEmail == "" || cfg.JiraAPIToken == "" || cfg.JiraDomain == "" {
		return TicketData{}, fmt.Errorf("missing Jira credentials (JIRA_EMAIL, JIRA_API_TOKEN, JIRA_DOMAIN)")
	}
	match := jiraKeyPattern.FindStringSubmatch(ticketURL)
	if match == nil {
		return TicketData{}, fmt.Errorf("invalid Jira ticket URL: %s", ticketURL)
	}
	ticketKey := match[1]
	apiURL := fmt.Sprintf("%s/rest/api/3/issue/%s", cfg.JiraDomain, ticketKey)

	var data struct {
		Key    string `json:"key"`
		Fields struct {
			Summary string `json:"summary"`
			Status  struct {
				Name string `json:"name"`
			} `json:"status"`
			Assignee struct {
				DisplayName string `json:"displayName"`
			} `json:"assignee"`
			Priority struct {
				Name string `json:"name"`
			} `json:"priority"`
			Project struct {
				Name string `json:"name"`
			} `json:"project"`
			Description any `json:"description"`
		} `json:"fields"`
	}

	err := retryWithBackoff(ctx, 3, func() error {
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, apiURL, nil)
		if err != nil {
			return err
		}
		req.SetBasicAuth(cfg.JiraEmail, cfg.JiraAPIToken)
		req.Header.Set("Accept", "application/json")
		return doJSON(req, &data)
	})
	if err != nil {
		return TicketData{}, err
	}

	return TicketData{
		Available: true, Source: "jira", Key: data.Key,
		Summary: data.Fields.Summary, Status: data.Fields.Status.Name,
		Priority:    data.Fields.Priority.Name,
		Description: strings.TrimSpace(parseADF(data.Fields.Description)),
		URL:         ticketURL,
	}, nil
}

func NewLinearTicketFetcherTool(cfg AppConfig) (agnt5.Tool, error) {
	return agnt5.NewTool("linear_ticket_fetcher", func(c context.Context, args map[string]any) (any, error) {
		ticketURL, _ := args["ticket_url"].(string)
		data, err := fetchLinearTicket(c, ticketURL, cfg)
		if err != nil {
			logError(c, "Linear fetch failed", "error", err)
			return nil, err
		}
		return data, nil
	},
		agnt5.WithToolDescription("Fetch Linear issue details using GraphQL API."),
		agnt5.WithToolSchema(map[string]any{
			"type": "object",
			"properties": map[string]any{
				"ticket_url": map[string]any{"type": "string", "description": "Full Linear ticket URL (e.g., https://linear.app/team/issue/PROJ-123)"},
			},
			"required": []string{"ticket_url"},
		}),
	)
}

func fetchLinearTicket(ctx context.Context, ticketURL string, cfg AppConfig) (TicketData, error) {
	if cfg.LinearKey == "" {
		return TicketData{}, fmt.Errorf("missing LINEAR_API_TOKEN in environment variables")
	}
	match := linearKeyPattern.FindStringSubmatch(ticketURL)
	if match == nil {
		return TicketData{}, fmt.Errorf("invalid Linear ticket URL: %s", ticketURL)
	}
	ticketKey := match[1]

	query := fmt.Sprintf(`{"query":"{ issue(id: \"%s\") { id identifier title description url state { name } assignee { name } team { name key } priority createdAt updatedAt dueDate } }"}`, ticketKey)

	var data struct {
		Data struct {
			Issue *struct {
				Identifier  string `json:"identifier"`
				Title       string `json:"title"`
				Description string `json:"description"`
				URL         string `json:"url"`
				State       struct {
					Name string `json:"name"`
				} `json:"state"`
				Assignee struct {
					Name string `json:"name"`
				} `json:"assignee"`
				Team struct {
					Name string `json:"name"`
				} `json:"team"`
				Priority float64 `json:"priority"`
			} `json:"issue"`
		} `json:"data"`
	}

	err := retryWithBackoff(ctx, 3, func() error {
		req, err := http.NewRequestWithContext(ctx, http.MethodPost, "https://api.linear.app/graphql", strings.NewReader(query))
		if err != nil {
			return err
		}
		req.Header.Set("Authorization", cfg.LinearKey)
		req.Header.Set("Content-Type", "application/json")
		return doJSON(req, &data)
	})
	if err != nil {
		return TicketData{}, err
	}
	if data.Data.Issue == nil {
		return TicketData{}, fmt.Errorf("no issue found for Linear key: %s", ticketKey)
	}
	issue := data.Data.Issue

	priority := ""
	if issue.Priority != 0 {
		priority = strconv.FormatFloat(issue.Priority, 'f', -1, 64)
	}

	return TicketData{
		Available: true, Source: "linear", Key: issue.Identifier,
		Summary: issue.Title, Status: issue.State.Name, Priority: priority,
		Description: issue.Description, URL: issue.URL,
	}, nil
}

func NewDetectTicketSourceTool() (agnt5.Tool, error) {
	return agnt5.NewTool("detect_ticket_source", func(c context.Context, args map[string]any) (any, error) {
		ticketURL, _ := args["ticket_url"].(string)
		switch {
		case strings.Contains(ticketURL, "atlassian.net"):
			logInfo(c, "Detected Jira ticket URL")
			return "jira", nil
		case strings.Contains(ticketURL, "linear.app"):
			logInfo(c, "Detected Linear ticket URL")
			return "linear", nil
		default:
			return nil, fmt.Errorf("unsupported ticketing platform URL: %s", ticketURL)
		}
	},
		agnt5.WithToolDescription("Determine ticketing platform (Jira or Linear) based on URL pattern."),
		agnt5.WithToolSchema(map[string]any{
			"type": "object",
			"properties": map[string]any{
				"ticket_url": map[string]any{"type": "string", "description": "Ticket URL to analyze"},
			},
			"required": []string{"ticket_url"},
		}),
	)
}

func logInfo(c context.Context, msg string, kv ...any) {
	if ctx, ok := c.(*agnt5.Context); ok {
		ctx.Logger().Info(msg, kv...)
	}
}

func logWarn(c context.Context, msg string, kv ...any) {
	if ctx, ok := c.(*agnt5.Context); ok {
		ctx.Logger().Warn(msg, kv...)
	}
}

func logError(c context.Context, msg string, kv ...any) {
	if ctx, ok := c.(*agnt5.Context); ok {
		ctx.Logger().Error(msg, kv...)
	}
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "..."
}
